tools-local/unix/search-tool.sh

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env bash
+
+# Searches for the specified tool in the environment path, and the path within the repository as specified in the .toolversions file.
+
+usage()
+{
+    echo "Usage: $0 ToolName StrictToolVersionMatch"
+    echo "  ToolName: Name of the tool to download."
+    echo "  StrictToolVersionMatch: A boolean indicating if the version of the tool to be searched should match the declared version."
+    echo "                          0 if no version check."
+    echo "                          1 if version should match the declared version."
+    echo "Searches for the tool in the environment path, and the path within the repository as specified in the .toolversions file."
+    echo "If search is successful then, returns the path to the tool."
+    echo "Exit 1 if search fails to find the tool."
+}
+
+if [ $# -ne 2 ]; then
+    usage
+    exit 1
+fi
+
+if [ -z "$1" ]; then
+    echo "Argument passed as ToolName is empty. Please provide a non-empty string."
+    exit 1
+fi
+
+if [ -z "$2" ]; then
+    echo "Argument passed as StrictToolVersionMatch is empty. Please provide a non-empty string."
+    exit 1
+fi
+
+toolName="$1"
+strictToolVersionMatch="$2"
+scriptPath="$(cd "$(dirname "$0")"; pwd -P)"
+. "$scriptPath/tool-helper.sh"
+probeLog="$scriptPath/probe-tool.log"
+declaredVersion="$(get_tool_config_value "$toolName" "DeclaredVersion")"
+
+
+# Displays the tool path, and exits the script.
+display_tool_path()
+{
+    echo "$toolPath"
+    echo "$(date) $toolName is available at $toolPath. Version is $toolVersion." >> "$probeLog"
+    exit
+}
+
+# Searches the tool in environment path.
+search_environment()
+{
+    echo "$(date) Searching for $toolName in environment path." >> "$probeLog"
+    hash "$toolName" 2>/dev/null
+
+    if [ $? -eq 0 ]; then
+        toolPath="$(which $toolName)"
+        toolVersion="$("$scriptPath/get-version.sh" "$toolName" "$toolPath")"
+
+        if [ "$strictToolVersionMatch" -eq "0" ]; then
+            # No strictToolVersionMatch. Hence, return the path found without further checks.
+            display_tool_path
+        else
+            # If strictToolVersionMatch is required then, ensure the version in environment path is same as declared version.
+            # If version matches then, return the path.
+            if [ "$toolVersion" == "$declaredVersion" ]; then
+                # Version available in environment path is the declared version.
+                display_tool_path
+            fi
+        fi
```

Comment:
I think this could be inverted and simplified:

```sh
if [ "$strictToolVersionMatch" -eq 1 ] && [ "$toolVersion" != "$declaredVersion" ]; then
    log "Version of $toolName at..."
    return 1
fi
echo "$toolPath"
log "$toolName is available at $toolPath. Version is $toolVersion."
```